src/ucs/debug/debug.c

```diff
@@ -536,26 +551,67 @@ ucs_status_t ucs_debug_lookup_address(void *address, ucs_debug_address_info_t *i
     return UCS_OK;
 }
 
-void ucs_debug_print_backtrace(FILE *stream, int strip)
+/**
+ * Create a backtrace from the calling location.
+ */
+void ucs_debug_backtrace_create(backtrace_h *bckt, int strip)
 {
-    char **symbols;
-    void *addresses[BACKTRACE_MAX];
-    int count, i, n;
+    size_t size = sizeof(**bckt);
+    ucs_status_t status;
 
-    fprintf(stream, "==== backtrace ====\n");
+    *bckt  = NULL;
+    status = ucs_mmap_alloc(&size, (void**)bckt, 0
+                            UCS_MEMTRACK_NAME("debug backtrace object"));
+    if (status != UCS_OK) {
+        ucs_assert(*bckt == NULL);
+        return;
+    }
 
-    count = backtrace(addresses, BACKTRACE_MAX);
-    symbols = backtrace_symbols(addresses, count);
-    n = 0;
-    for (i = strip; i < count; ++i) {
-        if (!ucs_debug_backtrace_is_excluded(addresses[i], symbols[i])) {
-            fprintf(stream, "   %2d  %s\n", n, symbols[i]);
-            ++n;
+    (*bckt)->size       = backtrace((*bckt)->addresses, BACKTRACE_MAX);
+    (*bckt)->symbols    = backtrace_symbols((*bckt)->addresses, (*bckt)->size);
+    (*bckt)->position   = strip;
+    (*bckt)->alloc_size = size;
+}
+
+/**
+ * Destroy a backtrace and free all memory.
+ *
+ * @param bckt          Backtrace object.
+ */
+void ucs_debug_backtrace_destroy(backtrace_h bckt)
+{
+    if (bckt != NULL) {
```

Comment:
why would it be NULL

Reply:
in case if no memory happened during backtrace creation(valgrind may force no memory) - then backtrace object is NULL and functions backtrace_next & backtrace_destroy just ignore such calls